mirgecom/profiling.py

```diff
@@ -0,0 +1,207 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from meshmode.array_context import PyOpenCLArrayContext
+import pyopencl as cl
+from pytools.py_codegen import PythonFunctionGenerator
+import loopy as lp
+import numpy as np
+from dataclasses import dataclass
+import pytools
+
+__doc__ = """
+.. autoclass:: PyOpenCLProfilingArrayContext
+"""
+
+
+@dataclass
+class ProfileResult:
+    """Class to hold the results of a single kernel execution."""
+    time: int
+    flops: int
+    bytes_accessed: int
+    footprint_bytes: int
+
+
+@dataclass
+class ProfileEvent:
+    """Class to hold a profile event that has not been seen by the profiler yet."""
+    cl_event: cl._cl.Event
+    program: lp.kernel.LoopKernel
+    kwargs: dict
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+    """An array context that profiles kernel executions.
+
+    .. automethod:: tabulate_profiling_data
+    .. automethod:: call_loopy
+
+    Inherits from :class:`meshmode.array_context.PyOpenCLArrayContext`."""
+
+    def __init__(self, queue, allocator=None) -> None:
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            raise RuntimeError("Profiling was not enabled in the command queue. "
+                 "Please create the queue with "
+                 "cl.command_queue_properties.PROFILING_ENABLE.")
+
+        self.profile_events = []
+        self.profile_results = {}
+        self.kernel_stats = {}
+
+    def _finish_profile_events(self) -> None:
+        # First, wait for event completion
+        if self.profile_events:
+            cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
+
+        for t in self.profile_events:
+            program = t.program
+            res = self._get_kernel_stats(program, t.kwargs)
+            res.time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            self.profile_results.setdefault(program, []).append(res)
+
+        self.profile_events = []
+
+    def tabulate_profiling_data(self) -> pytools.Table:
+        """Returns a :class:`pytools.Table` with the profiling results."""
+        self._finish_profile_events()
+
+        tbl = pytools.Table()
+
+        tbl.add_row(["Function", "Calls", "Time_min [s]", "Time_avg [s]",
+            "Time_max [s]", "GFlops_min", "GFlops_avg", "GFlops_max", "GMemAcc_min",
+            "GMemAcc_avg", "GMemAcc_max", "Bandwidth [GByte/s]",
+            "Footprint [GByte]"])
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v.time / 1e9 for v in value]
+            flops = [v.flops / 1e9 for v in value]
+            bytes_accessed = [v.bytes_accessed / 1e9 for v in value]
+            footprint_bytes = [v.footprint_bytes / 1e9
+                if v.footprint_bytes is not None else 0 for v in value]
```

Comment:
This will throw off your min, max, mean. "0" is a really poor thing to use for "invalid value".

Reply:
Fixed by only storing the non-None values (and printing "n/a" as the mean if there aren't any values at all)